GithubBrowserSample/app/src/main/java/com/android/example/github/repository/CoroutineNetworkBoundResource.kt

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2019 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.example.github.repository
+
+import androidx.annotation.MainThread
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.LiveDataScope
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.distinctUntilChanged
+import androidx.lifecycle.liveData
+import androidx.lifecycle.map
+import com.android.example.github.api.ApiEmptyResponse
+import com.android.example.github.api.ApiErrorResponse
+import com.android.example.github.api.ApiResponse
+import com.android.example.github.api.ApiSuccessResponse
+import com.android.example.github.vo.Resource
+import com.android.example.github.vo.Status
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import java.util.concurrent.CancellationException
+
+
+fun <ResultType, RequestType> networkBoundResource(
+    saveCallResult: suspend (RequestType) -> Unit,
```

Comment:
i'm not 100% sure whether i want the class type api or function type API (old one vs new one).

This one i think reads nicer vs the other one gets better IDE support?
also this one has a little advantage of inferring all types from fetch and loadFromDb methods.

Reply:
actually i think i now like it better. Type inference make wonders.